uring/src/main/scala/fs2/io/uring/net/UringDatagramSocket.scala

```diff
@@ -0,0 +1,94 @@
+package fs2
+package io
+package uring
+package net
+
+import cats.effect.kernel.Async
+import cats.effect.kernel.Resource
+import cats.effect.kernel.Sync
+import cats.effect.std.Mutex
+import cats.syntax.all._
+import com.comcast.ip4s.IpAddress
+import com.comcast.ip4s.SocketAddress
+import fs2.Pipe
+import fs2.io.net.Datagram
+import fs2.io.net.DatagramSocket
+import fs2.io.uring.unsafe.uring._
+import fs2.io.uring.unsafe.util._
+
+import scala.scalanative.libc.errno._
+import scala.scalanative.posix.sys.socket._
+import scala.scalanative.unsafe._
+import scala.scalanative.unsigned._
+import com.comcast.ip4s.MulticastJoin
+
+private[net] final class UringDatagramSocket[F[_]](
+    ring: Uring[F],
+    fd: Int,
+    buffer: NonResizableBuffer[F],
+    defaultReadSize: Int,
+    readMutex: Mutex[F],
+    writeMutex: Mutex[F]
+)(implicit F: Async[F])
+    extends DatagramSocket[F] {
+
+  private[this] def recvmsg(msg: Ptr[msghdr], flags: Int): F[Int] =
+    ring.call(io_uring_prep_recvmsg(_, fd, msg, flags))
```

Comment:
So the `msghdr` struct is defined like this in Scala Native:

```scala
 type msghdr = CStruct7[
    Ptr[Byte], // msg_name
    socklen_t, // msg_namelen
    Ptr[uio.iovec], // msg_iov
    CInt, // msg_iovlen
    Ptr[Byte], // msg_control
    socklen_t, // msg_crontrollen
    CInt // msg_flags
  ]
```
https://github.com/scala-native/scala-native/blob/364885493c7be0507fd40e62ad484dde995c352b/posixlib/src/main/scala/scala/scalanative/posix/sys/socket.scala#L23-L31

which is equivalent to this C declaration:
```C
struct iovec {                    /* Scatter/gather array items */
    void  *iov_base;              /* Starting address */
    size_t iov_len;               /* Number of bytes to transfer */
};

struct msghdr {
    void         *msg_name;       /* optional address */
    socklen_t     msg_namelen;    /* size of address */
    struct iovec *msg_iov;        /* scatter/gather array */
    size_t        msg_iovlen;     /* # elements in msg_iov */
    void         *msg_control;    /* ancillary data, see below */
    size_t        msg_controllen; /* ancillary data buffer len */
    int           msg_flags;      /* flags on received message */
};
```
https://linux.die.net/man/2/recvmsg

So I think what we want to do is:

1. before calling `io_uring_prep_recvmsg`, we should setup `msghdr` with the pre-allocated buffer. I think we should provide it as the `msg_iov`.

2. after `io_uring_prep_recvmsg` completes, we can read the datagram out of the buffer. The return value will tell us how many bytes were read. Additionally, we can retrieve the address out of `msg_name` using `msg_namelen` as the length.

Reply:
You have clarified all my doubts, thank you very much :)